nixos/image: systemd-repart based GPT image builder

nikstur opened the pull request:
###### Description of changes

Add a new module to build images via `systemd-repart`. This module exposes a shallow wrapper around `systemd-repart` that can be used to assemble appliance images by populating the Efi System Partition (ESP) yourself.

The eventual ambitions for this image builder are much larger though:

- Have the ability to install all supported boot loaders (by assembling the ESP beforehand and simply copying the contents into the image)
- Expose the builder as a snippet, to optionally build disk images just-in-time (so they are not stored in the Nix store and thus not cached)
- Replace `make-disk-image.nix` for all GPT/UEFI use cases (including as the default image creator for qemu-vm.nix)

Generally, this module wants to remain a shallow wrapper around `systemd-repart` and all additional funcitonality should be implemented outside of this module (e.g. populating an ESP) and then just copy into the image via the provided mechanisms.

I'm not sure where the docs should go. I feel they are too early in the manual. Also these are not really reference material, so I'm not sure if they even belong in the manual. Maybe nix.dev is a better place?

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nikstur committed:
nixos/tests/appliance-repart-image: init

nikstur committed:
nixos/image: add repart builder

RaitoBezarius commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,199 @@
+  };
+in
+{
+  options.image = {
```
Not a fan of taking the namespace like this.

RaitoBezarius replied:
This should make it clear this is using systemd-repart somehow.

nikstur replied:
Mhmm does it need to encode systemd-repart though? Seems like a leaky abstraction that is not ncessary. But I can see that `options.image` is very fundamental. Would `options.diskImage` be better or what do you have in mind?

RaitoBezarius replied:
I mean, if we do have other ways to build images, you will have to accommodate for them.
So except if you are ready to accept complicated discussions to compromise on the structure of this module, I don't think it's a good idea to have a global `image` module.

It would make more sense for me to have `systemd.repartImage.xyz`, it makes it clear this is a repart-based abstraction. The leaky abstraction is necessary because you have `repartConfig` in the whole thing.

Probably, we could have a `image.backend = "repart";` and `image.*` which have sensible options for general purpose and a proper abstraction, but I feel like it's premature?

nikstur replied:
So, what about  `options.image.repart`? That leaves plenty of space for other backends and my current abstraction is leaky anyways already because it exposes `repartConfig`.

Yeah, you're right, let's keep this specific instead of generic for now.

RaitoBezarius replied:
`image.repart` sounds good to me. :)

RaitoBezarius commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,199 @@
+      storePaths = lib.mkOption {
+        type = with lib.types; listOf path;
+        default = [ ];
+        description = lib.mdDoc "The store paths to inlcude in the partition.";
```
```suggestion
        description = lib.mdDoc "The store paths to include in the partition.";
```

RaitoBezarius commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,199 @@
+        description = lib.mdDoc "The store paths to inlcude in the partition.";
+      };
+
+      stripStorePaths = lib.mkOption {
```
```suggestion
      stripNixStorePrefix = lib.mkOption {
```

The name is very unclear IMHO.

RaitoBezarius commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,199 @@
+
+    seed = lib.mkOption {
+      type = with lib.types; nullOr str;
+      default = "0867da16-f251-457d-a9e8-c31f9a3c220b";
```
How is this UUID chosen? Please add a comment.

RaitoBezarius commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,199 @@
+
+        makeClosure = paths: pkgs.closureInfo { rootPaths = paths; };
+
+        addClosure = _name: partitionConfig: partitionConfig // (
```
Add a comment to explain what is this supposed to do.

RaitoBezarius commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,199 @@
+          (f: f != null)
+          (lib.mapAttrsToList (_n: v: v.repartConfig.Format or null) cfg.partitions);
+
+        fileSystemTools = builtins.concatMap (f: fileSystemToolMapping."${f}") fileSystems;
```
```suggestion
        requiredFileSystemTools = builtins.concatMap (f: fileSystemToolMapping."${f}") fileSystems;
```

nikstur replied:
Is this Java? I don't think adding `required` makes this any clearer.

RaitoBezarius replied:
:DDDDDDDDDDDDDDDDDD

RaitoBezarius replied:
I mean, this is the filesystem tools *required* to manipulate the filesystems, so I do think it's clearer, but it's fine, this is a nit.

RaitoBezarius commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,199 @@
+
+        addClosure = _name: partitionConfig: partitionConfig // (
+          lib.optionalAttrs
+            (partitionConfig ? "storePaths" && partitionConfig.storePaths != [ ])
```
```suggestion
            (partitionConfig.storePaths or []) != [ ]
```

RaitoBezarius commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,199 @@
+        finalPartitions = lib.mapAttrs addClosure cfg.partitions;
+
+
+        amendRepartDefinitions = pkgs.runCommand "amend-repart-definitions.py"
```
Don't we have a writePythonScript that does the black ruff mypy stuff for you?

nikstur replied:
Not aware of that and can't find it.

RaitoBezarius replied:
```
  # writePython3 takes a name an attributeset with libraries and some python3 sourcecode and
  # returns an executable
  #
  # Example:
  # writePython3 "test_python3" { libraries = [ pkgs.python3Packages.pyyaml ]; } ''
  #   import yaml
  #
  #   y = yaml.load("""
  #     - test: success
  #   """)
  #   print(y[0]['test'])
  # ''
```

in pkgs/build-support/writers

But it's not super good for Python I see.

RaitoBezarius commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,199 @@
+        mkdir -p $out
+        cd $out
+
+        fakeroot systemd-repart \
```
nit: Do you need fakeroot or can you achieve this with `unshare --map-root-user` ?

nikstur replied:
No idea what `unshare` does but I'll give it a try.

RaitoBezarius replied:
Unshare just spawn a new user namespace unprivileged with root UID mapped inside of it.

RaitoBezarius replied:
I'd expect systemd-repart to work unprivileged with this type of stuff given the folks who are behind it :P.

nikstur replied:
I can't get it to work with unshare. I'll put it on the future work list.

RaitoBezarius commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,199 @@
+        {
+          nativeBuildInputs = with pkgs; [
+            fakeroot
+            systemdMinimal
```
Does `systemdMinimal` have `systemd-repart`? This will change in the future IMHO? Let's just have the proper `systemd` closure, we will see later how to reduce this if needed.

nikstur replied:
Fair. It's a build time dependency anyways.

RaitoBezarius commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,199 @@
+          image.raw
+      '';
+
+    meta.maintainers = with lib.maintainers; [ nikstur ];
```
nit: I'm torn on the fact you made a building image chapter, I feel like this module should have its own documentation locally with `doc = ./building-images-with-systemd-repart.md` and we can have a global documentation referencing our options for that.

RaitoBezarius replied:
Ah I see, it answers your questions on the manual page ;).

nikstur replied:
Do you have a link?

RaitoBezarius replied:
https://github.com/NixOS/nixpkgs/tree/master/nixos/modules/system/boot/loader/external

RaitoBezarius commented on nixos/tests/appliance-gpt-image.nix:
```diff
@@ -0,0 +1,114 @@
+{ lib, ... }:
+
+let
+  rootParitionlabel = "root";
```
```suggestion
  rootPartitionlabel = "root";
```

RaitoBezarius commented on nixos/tests/appliance-gpt-image.nix:
```diff
@@ -0,0 +1,114 @@
+    virtualisation.useEFIBoot = true;
+
+    # Disable boot loaders because we install one "manually".
+    boot.loader.grub.enable = false;
```
In the future, a GPT image appliance will be able to provide bootloaders via #244907 by having a `appliance-loader` or you can add compatibility with that. Therefore, we will still be able to validate that a bootloader is contained inside the appliance.

nikstur replied:
Cool, let's remove this once https://github.com/NixOS/nixpkgs/pull/244907 lands.

RaitoBezarius replied:
Just add a `# TODO(raitobezarius): revisit this when #244907 lands`

RaitoBezarius commented on nixos/tests/appliance-gpt-image.nix:
```diff
@@ -0,0 +1,114 @@
+
+    virtualisation.fileSystems = lib.mkForce {
+      "/" = {
+        device = "/dev/disk/by-partlabel/${rootParitionlabel}";
```
```suggestion
        device = "/dev/disk/by-partlabel/${rootPartitionlabel}";
```

RaitoBezarius replied:
Don't you just need to override `virtualisation.rootDevice` here?

nikstur replied:
I believe this would make this less readable and more complex. The right way to use this image is via the changes from https://github.com/NixOS/nixpkgs/pull/238847

RaitoBezarius replied:
Then, I disagree, I feel like you should do `filesystems."/" = mkImageMediaOverride { ... }` and ensure that this works out of the box in virtualization because `mkImageMediaOverride` should be of higher priority.
Otherwise, we can consider a higher priority override.

Then, your changes of #238847 can become simpler I believe.

RaitoBezarius replied:
This is the same issue as the netboot media and the ISO media.

RaitoBezarius commented on nixos/tests/appliance-gpt-image.nix:
```diff
@@ -0,0 +1,114 @@
+              "/EFI/BOOT/BOOT${lib.toUpper efiArch}.EFI".source =
+                "${pkgs.systemd}/lib/systemd/boot/efi/systemd-boot${efiArch}.efi";
+
+              "${bootLoaderConfigPath}".source = pkgs.writeText "nixos.conf" ''
```
systemd-boot renderer should be extracted and made available as some sort of `pkgs.writeBLS` or something I guess.

nikstur replied:
A simple wrapper around a structural setting for boot loader configs would be nice. A time for this is if/when we add another abstraction for building appliances.

RaitoBezarius replied:
Yeah, a TODO is good for this one.

RaitoBezarius reviewed (changes requested):
Very good! My biggest beef is the namespacing. :)

arianvp reviewed (commented):
I think it would be better to have a derivation that takes a bootspec as input and outputs repart definitions as output instead of having a DSL at the nix expression Level for setting partition contents.

This will also make it easier to have this as a standalone script that builds images outside of the nix store and makes the builder usable outside of NixOS

arianvp commented:
Would be nice to add or document that you can run 

```
      ${config.nix.package}/bin/nix-store --load-db < /nix/store/nix-path-registration

```

From `closureInfo`

To get a working NixOS (non appliance) image if nix binary is in the target image

Like here:

https://github.com/arianvp/server-optimised-nixos/blob/9e3de35bb5885c22695b953976466aa3c15ac8cc/modules/base.nix#L39

arianvp commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,202 @@
+            systemd
+          ] ++ fileSystemTools;
+        } ''
+        amendedRepartDefinitions=$(${amendRepartDefinitions} ${partitions} ${definitionsDirectory})
```
I think we should allow partition definitions that are generated by derivations as well. This allows us to generate a partition from a `bootspec` file for example. (e.g. populate systemd boot-loader entries in the `esp` partition using metadata from `boot.json` in the `toplevel` closure`)

arianvp replied:
Ah on further inspection is already works as expected.

Can just set `.source = "${derivation here}`

nikstur replied:
Yes you can write arbitary files to arbitrary locations inside arbitrary partitions. If you want to use the boot.json to assemble something thats already posssible.

You could even generate the actual partiton definitions in a derivation and then just import from derivation. Depending on your use-case this might be fine.

arianvp commented on nixos/modules/image/gpt.nix:
```diff
@@ -0,0 +1,202 @@
+        addClosure = _name: partitionConfig: partitionConfig // (
+          lib.optionalAttrs
+            (partitionConfig.storePaths or [ ] != [ ])
+            { closure = "${makeClosure partitionConfig.storePaths}/store-paths"; }
```
```suggestion
            { closure = "${makeClosure partitionConfig.storePaths}"; }
```

so that we have access to `$out/registration` in the python script

nikstur replied:
See my other comments.

arianvp commented on nixos/modules/image/amend-repart-definitions.py:
```diff
@@ -0,0 +1,113 @@
+        return
+
+    copy_files_lines: list[str] = []
+    with open(closure, "r") as f:
```
```suggestion
    nix_path_registration_source = closure+"/registration"
    nix_path_registration_target = "nix-path-registration" if strip_nix_store_prefix else "/nix/store/nix-path-registration"
    copy_file_lines.append(f"CopyFiles={nix_path_registration_source}:{nix_path_registration_target}")
    with open(closure+"/store-paths", "r") as f:
```

nikstur replied:
Your suggested solution has the downside that it puts the registration into the nix store. We don't actually have to touch the python script to achieve get the registration into our image though.

You can just do `contents."/registration".source = "${closure}/registration"`. Because the closure is a derivation, this would just use the cached derivation and not actually compute anything new. 

Now you can also pick where the registration ends up in your image. E.g. it could be on the root partition while the nix store (potentially dm-verity backed) is on a completely different partition.

On a side note: I'd really want to avoid to proliferate `boot.postBootCommands` in Nixpkgs because I'd love to see scripted activation be replaced by parallel services that perform the activation. Of course you could achieve the same if you just write a service yourself.

arianvp replied:
Cool! Clear! I might make a follow-up PR to expose this as an option then (and try to come up with something different than `postBootCommands`).

nikstur committed:
nixos/image: add repart builder docs

nikstur committed:
nixos/lib: systemd definition files function

Add a re-usable function that converts an attrset to a directory
containing systemd definition files.

nikstur commented:
> takes a bootspec as input and outputs repart definitions as output instead of having a DSL at the nix expression Level for setting partition contents.

The bootspec does not contain all the information necessary to build an image. Additionally, if we statically assemble an image, and not rely on the generation mechanism, bootspec doesn't really benefit us. We can just use the normal templating mechanisms Nix provides. The issue bootspec really addresses (at least to me) is that it has all the information in place for a bootloader installer that cannot rely on normal templating in Nix. 

> makes the builder usable outside of NixOS

Outside NixOS, you should probably just use systemd-repart directly or use mkosi wich also uses systemd-repart under the hood.

> To get a working NixOS (non appliance) image if nix binary is in the target image

I want to address this outside of this PR. This PR really just adds the NixOS primitives to create an image with systemd-repart. I believe this is also a more elegan approach. make-disk-image.nix for example suffers from a lot of complexity and very few abstractions that make your life easier. 

The way I envision to eventually create a fully working Nix image (that also supports generations and can be used by qemu-vm.nix) is by creating all the necessary pieces in separate derivations (or at least outside the image builder itself) and just copy the resulting files to the right spots inside the image via the primites exposed in `systemd.repart`. So we would for example have an ESP builder that can assemble the directory tree for an entire ESP (with generations) and then just tell system-repart to copy this tree into the ESP.  Registration of the closure is then done via a systemd service in userspace.

nikstur commented:
@ofborg test systemd-sysupdate systemd-repart.basic appliance-repart-image
